File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/seed/FakeHttpsUrlConnection.java
package io.github.pr0methean.betterrandom.seed;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Proxy;
import java.net.URL;
import java.security.cert.Certificate;
import javax.annotation.Nullable;
import javax.net.ssl.HttpsURLConnection;

/**
 * Used for testing {@link RandomDotOrgSeedGenerator}.
 */
public class FakeHttpsUrlConnection extends HttpsURLConnection {

  private static final Certificate[] CERTIFICATES = new Certificate[0];
  @Nullable private final Proxy proxy;
  private volatile boolean disconnected = false;
  private final ByteArrayOutputStream os = new ByteArrayOutputStream();
  private final ByteArrayInputStream is;

  public FakeHttpsUrlConnection(final URL url, @Nullable final Proxy proxy,
      final byte[] responseBody) {
    super(url);
    this.proxy = proxy;
    is = new ByteArrayInputStream(responseBody);
  }

  @Override public InputStream getInputStream() {
    if (disconnected) {
      throw new IllegalStateException("Already disconnected");
    }
    return is;
  }

  @Override public OutputStream getOutputStream() {
    return os;
  }

  @Override public String getCipherSuite() {
    return null;
  }

  @Override public Certificate[] getLocalCertificates() {
    return CERTIFICATES;
  }

  @Override public Certificate[] getServerCertificates() {
    return CERTIFICATES;
  }

  @Override public void disconnect() {
    disconnected = true;
  }

  @Override public boolean usingProxy() {
    return (proxy != null) && !proxy.equals(Proxy.NO_PROXY);
  }

  @Override public void connect() {
    connected = true;
  }

  @Override protected void finalize() {
    try {
      if (is != null) {
        is.close();
      }
      os.close();
    } catch (final IOException ignored) {}
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/seed/FakeHttpsUrlConnection.kt
package io.github.pr0methean.betterrandom.seed

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.Proxy
import java.net.URL
import java.security.cert.Certificate
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLPeerUnverifiedException
import org.slf4j.LoggerFactory

/**
 * Used for testing [RandomDotOrgSeedGenerator].
 */
class FakeHttpsUrlConnection(url: URL, val proxy: Proxy?, responseBody: ByteArray) : HttpsURLConnection(url) {
    @Volatile
    var isDisconnected = false
        private set
    private val os = ByteArrayOutputStream()
    private val `is`: ByteArrayInputStream?

    val requestBody: ByteArray
        get() = os.toByteArray()

    init {
        `is` = ByteArrayInputStream(responseBody)
    }

    @Throws(IOException::class)
    override fun getInputStream(): InputStream? {
        if (isDisconnected) {
            throw IllegalStateException("Already disconnected")
        }
        return `is`
    }

    @Throws(IOException::class)
    override fun getOutputStream(): OutputStream {
        return os
    }

    override fun getCipherSuite(): String? {
        return null
    }

    override fun getLocalCertificates(): Array<Certificate> {
        return CERTIFICATES
    }

    @Throws(SSLPeerUnverifiedException::class)
    override fun getServerCertificates(): Array<Certificate> {
        return CERTIFICATES
    }

    override fun disconnect() {
        isDisconnected = true
    }

    override fun usingProxy(): Boolean {
        return proxy != null && proxy != Proxy.NO_PROXY
    }

    @Throws(IOException::class)
    override fun connect() {
        connected = true
    }

    protected override fun finalize() {
        try {
            `is`?.close()
            os.close()
        } catch (e: IOException) {
            LoggerFactory.getLogger(FakeHttpsUrlConnection::class.java).error("Failed to close streams", e)
        }

    }

    companion object {

        private val CERTIFICATES = arrayOfNulls<Certificate>(0)
    }
}
